File: ios/MRT/Features/Chat/ChatViewModel.swift
import Foundation
import SwiftUI

struct FeatureChatMessage: Identifiable, Equatable {
    enum Role: Equatable {
        case user
        case assistant
        case system
    }

    let id: UUID
    let sessionID: String?
    var content: String
    var isComplete: Bool
    let role: Role
    let timestamp: Date

    init(
        id: UUID = UUID(),
        sessionID: String?,
        content: String,
        isComplete: Bool,
        role: Role,
        timestamp: Date = Date()
    ) {
        self.id = id
        self.sessionID = sessionID
        self.content = content
        self.isComplete = isComplete
        self.role = role
        self.timestamp = timestamp
    }

    var timeAgo: String {
        let formatter = RelativeDateTimeFormatter()
        formatter.unitsStyle = .abbreviated
        return formatter.localizedString(for: timestamp, relativeTo: Date())
    }
}

@MainActor
final class ChatViewModel: ObservableObject {
    private struct ConnectionConfiguration: Equatable {
        let host: String
        let port: Int
        let mode: ConnectionMode
    }

    @Published var messages: [FeatureChatMessage] = []
    @Published var inputText = ""
    @Published var connectionState: ConnectionState
    @Published var pendingApproval: Mrt_ApprovalRequest? {
        didSet {
            if pendingApproval != nil {
                connectionState = .showingApproval
            } else if connectionState == .showingApproval {
                connectionState = .connected
            }
        }
    }
    @Published var activeSessionID: String? {
        didSet {
            rebuildMessages()
        }
    }

    private let connectionManager: ConnectionManaging
    private var localMessages: [FeatureChatMessage] = []
    private var remoteMessages: [ChatMessage] = []
    private var remoteMessageTimestamps: [UUID: Date] = [:]
    private var lastConnectionConfiguration: ConnectionConfiguration?

    init(connectionManager: ConnectionManaging) {
        self.connectionManager = connectionManager
        self.connectionState = connectionManager.state
        self.remoteMessages = connectionManager.messages
        self.pendingApproval = connectionManager.pendingApproval
        bindConnectionManager()
        rebuildMessages()
    }

    var isLoading: Bool {
        get { connectionState == .loading }
        set {
            if newValue {
                connectionState = .loading
            } else if connectionState == .loading {
                connectionState = .connected
            }
        }
    }

    func sendPrompt() async {
        let prompt = inputText.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !prompt.isEmpty else { return }
        guard let activeSessionID else { return }

        let userMessage = FeatureChatMessage(
            sessionID: activeSessionID,
            content: prompt,
            isComplete: true,
            role: .user
        )
        localMessages.append(userMessage)
        rebuildMessages()

        inputText = ""
        connectionState = .loading

        do {
            try await connectionManager.sendPrompt(prompt, sessionID: activeSessionID)
        } catch {
            localMessages.append(
                FeatureChatMessage(
                    sessionID: activeSessionID,
                    content: "Unable to send prompt right now.",
                    isComplete: true,
                    role: .system
                )
            )
            rebuildMessages()
            connectionState = connectionManager.state
        }
    }

    func respondToApproval(_ approved: Bool) async {
        guard let approvalID = pendingApproval?.approvalID else { return }

        do {
            try await connectionManager.respondToApproval(approvalID, approved: approved)
            pendingApproval = nil
        } catch {
        }
    }

    var lastMessageSignature: String {
        guard let lastMessage = messages.last else { return "empty" }
        return "\(lastMessage.id.uuidString):\(lastMessage.content.count):\(lastMessage.isComplete)"
    }

    func connectIfNeeded(host: String, port: Int, mode: ConnectionMode) async {
        let configuration = ConnectionConfiguration(
            host: host.trimmingCharacters(in: .whitespacesAndNewlines),
            port: port,
            mode: mode
        )
        let currentState = connectionManager.state
        let canRetryCurrentConfiguration = currentState == .disconnected || currentState == .reconnecting
        guard configuration != lastConnectionConfiguration || canRetryCurrentConfiguration else { return }
        lastConnectionConfiguration = configuration
        guard mode == .direct else { return }

        do {
            try await connectionManager.connect(host: configuration.host, port: configuration.port)
        } catch {
            connectionState = connectionManager.state
        }
    }

    private func bindConnectionManager() {
        connectionManager.onStateChange = { [weak self] newState in
            Task { @MainActor in
                self?.connectionState = newState
            }
        }
        connectionManager.onMessagesChange = { [weak self] newMessages in
            Task { @MainActor in
                self?.remoteMessages = newMessages
                self?.cacheRemoteMessageTimestamps(for: newMessages)
                self?.rebuildMessages()
            }
        }
        connectionManager.onPendingApprovalChange = { [weak self] approval in
            Task { @MainActor in
                self?.pendingApproval = approval
            }
        }
    }

    private func rebuildMessages() {
        let mappedRemote = remoteMessages.map { message in
            FeatureChatMessage(
                id: message.id,
                sessionID: message.sessionID,
                content: message.content,
                isComplete: message.isComplete,
                role: mapRole(message.role),
                timestamp: remoteMessageTimestamps[message.id] ?? Date()
            )
        }
        messages = (localMessages + mappedRemote)
            .filter(isVisibleInActiveThread)
            .enumerated()
            .sorted { lhs, rhs in
                if lhs.element.timestamp == rhs.element.timestamp {
                    return lhs.offset < rhs.offset
                }
                return lhs.element.timestamp < rhs.element.timestamp
            }
            .map(\.element)
    }

    private func mapRole(_ role: ChatMessage.Role) -> FeatureChatMessage.Role {
        switch role {
        case .assistant:
            return .assistant
        case .system:
            return .system
        }
    }

    private func isVisibleInActiveThread(_ message: FeatureChatMessage) -> Bool {
        if message.sessionID == activeSessionID {
            return true
        }

        return message.sessionID == nil && message.role == .system
    }

    private func cacheRemoteMessageTimestamps(for messages: [ChatMessage]) {
        var nextTimestamp = Date()
        for message in messages where remoteMessageTimestamps[message.id] == nil {
            remoteMessageTimestamps[message.id] = nextTimestamp
            nextTimestamp = nextTimestamp.addingTimeInterval(0.001)
        }
    }
}

File: android/app/src/main/java/com/mrt/app/features/chat/ChatViewModel.kt
package com.mrt.app.features.chat

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import com.mrt.app.core.models.ChatMessage
import com.mrt.app.core.network.ConnectionManaging
import com.mrt.app.core.network.ConnectionState
import com.mrt.app.core.storage.ConnectionMode
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.CoroutineStart
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.flow.collectLatest
import kotlinx.coroutines.launch
import mrt.Mrt

class ChatViewModel(
    private val connectionManager: ConnectionManaging,
    private val scope: CoroutineScope = CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate),
    private val nowMs: () -> Long = { System.currentTimeMillis() },
) {
    data class FeatureChatMessage(
        val id: String,
        val sessionId: String?,
        val content: String,
        val isComplete: Boolean,
        val role: Role,
        val timestampMs: Long,
        val order: Long,
    ) {
        enum class Role {
            USER,
            ASSISTANT,
            SYSTEM,
        }
    }

    private data class ConnectionConfiguration(
        val host: String,
        val port: Int,
        val mode: ConnectionMode,
    )

    var messages by mutableStateOf<List<FeatureChatMessage>>(emptyList())
        private set
    var inputText by mutableStateOf("")
    var connectionState by mutableStateOf(connectionManager.state.value)
    var pendingApproval by mutableStateOf<Mrt.ApprovalRequest?>(connectionManager.pendingApproval.value)
        private set
    private var activeSessionIdState by mutableStateOf<String?>(null)
    var activeSessionId: String?
        get() = activeSessionIdState
        set(value) {
            activeSessionIdState = value
            rebuildMessages()
        }

    private var localMessages: List<FeatureChatMessage> = emptyList()
    private var remoteMessages: List<ChatMessage> = connectionManager.messages.value
    private val remoteMessageOrders = linkedMapOf<String, Long>()
    private var nextOrder = 0L
    private var lastConnectionConfiguration: ConnectionConfiguration? = null

    init {
        scope.launch(context = Dispatchers.Unconfined, start = CoroutineStart.UNDISPATCHED) {
            connectionManager.state.collectLatest { state ->
                connectionState = state
            }
        }
        scope.launch(context = Dispatchers.Unconfined, start = CoroutineStart.UNDISPATCHED) {
            connectionManager.messages.collectLatest { messages ->
                remoteMessages = messages
                cacheRemoteMessageOrdering(messages)
                rebuildMessages()
            }
        }
        scope.launch(context = Dispatchers.Unconfined, start = CoroutineStart.UNDISPATCHED) {
            connectionManager.pendingApproval.collectLatest { approval ->
                pendingApproval = approval
                if (approval != null) {
                    connectionState = ConnectionState.SHOWING_APPROVAL
                } else if (connectionState == ConnectionState.SHOWING_APPROVAL) {
                    connectionState = ConnectionState.CONNECTED
                }
            }
        }
        rebuildMessages()
    }

    var isLoading: Boolean
        get() = connectionState == ConnectionState.LOADING
        set(value) {
            if (value) {
                connectionState = ConnectionState.LOADING
            } else if (connectionState == ConnectionState.LOADING) {
                connectionState = ConnectionState.CONNECTED
            }
        }

    val canSendPrompt: Boolean
        get() = !isLoading && activeSessionId != null && inputText.trim().isNotEmpty()

    val inputAssistiveMessage: String?
        get() = if (activeSessionId == null) {
            "Select or create a session to start chatting"
        } else {
            null
        }

    val sendButtonLabel: String
        get() = if (isLoading) "Sending" else "Send"

    val lastMessageSignature: String
        get() = messages.lastOrNull()?.let { "${it.id}:${it.content.length}:${it.isComplete}" } ?: "empty"

    suspend fun connectIfNeeded(host: String, port: Int, mode: ConnectionMode) {
        val configuration = ConnectionConfiguration(
            host = host.trim(),
            port = port,
            mode = mode,
        )
        val currentState = connectionManager.state.value
        val canRetryCurrentConfiguration =
            currentState == ConnectionState.DISCONNECTED || currentState == ConnectionState.RECONNECTING

        if (configuration == lastConnectionConfiguration && !canRetryCurrentConfiguration) {
            return
        }
        lastConnectionConfiguration = configuration
        if (mode != ConnectionMode.DIRECT) {
            connectionManager.disconnect()
            connectionState = connectionManager.state.value
            return
        }

        try {
            connectionManager.connect(host = configuration.host, port = configuration.port)
        } catch (_: Throwable) {
            connectionState = connectionManager.state.value
        }
    }

    suspend fun sendPrompt() {
        val prompt = inputText.trim()
        val sessionId = activeSessionId
        if (prompt.isEmpty() || sessionId == null) {
            return
        }

        localMessages = localMessages + FeatureChatMessage(
            id = "local-${nextOrder + 1}",
            sessionId = sessionId,
            content = prompt,
            isComplete = true,
            role = FeatureChatMessage.Role.USER,
            timestampMs = nowMs(),
            order = nextOrder(),
        )
        rebuildMessages()

        inputText = ""
        connectionState = ConnectionState.LOADING

        try {
            connectionManager.sendPrompt(prompt = prompt, sessionId = sessionId)
        } catch (_: Throwable) {
            localMessages = localMessages + FeatureChatMessage(
                id = "local-${nextOrder + 1}",
                sessionId = sessionId,
                content = "Unable to send prompt right now.",
                isComplete = true,
                role = FeatureChatMessage.Role.SYSTEM,
                timestampMs = nowMs(),
                order = nextOrder(),
            )
            rebuildMessages()
            connectionState = connectionManager.state.value
        }
    }

    suspend fun respondToApproval(approved: Boolean) {
        val approvalId = pendingApproval?.approvalId ?: return
        try {
            connectionManager.respondToApproval(approvalId = approvalId, approved = approved)
            pendingApproval = null
            if (connectionState == ConnectionState.SHOWING_APPROVAL) {
                connectionState = ConnectionState.CONNECTED
            }
        } catch (_: Throwable) {
        }
    }

    private fun rebuildMessages() {
        val mappedRemote = remoteMessages.map { message ->
            FeatureChatMessage(
                id = message.id.toString(),
                sessionId = message.sessionId,
                content = message.content,
                isComplete = message.isComplete,
                role = when (message.role) {
                    ChatMessage.Role.ASSISTANT -> FeatureChatMessage.Role.ASSISTANT
                    ChatMessage.Role.SYSTEM -> FeatureChatMessage.Role.SYSTEM
                },
                timestampMs = remoteMessageOrders[message.id.toString()] ?: nowMs(),
                order = remoteMessageOrders[message.id.toString()] ?: 0L,
            )
        }
        messages = (localMessages + mappedRemote)
            .filter(::isVisibleInActiveThread)
            .sortedWith(compareBy<FeatureChatMessage> { it.order }.thenBy { it.timestampMs })
    }

    private fun isVisibleInActiveThread(message: FeatureChatMessage): Boolean {
        if (message.sessionId == activeSessionId) {
            return true
        }
        return message.sessionId == null && message.role == FeatureChatMessage.Role.SYSTEM
    }

    private fun cacheRemoteMessageOrdering(messages: List<ChatMessage>) {
        messages.forEach { message ->
            val key = message.id.toString()
            if (remoteMessageOrders[key] == null) {
                remoteMessageOrders[key] = nextOrder()
            }
        }
    }

    private fun nextOrder(): Long {
        nextOrder += 1
        return nextOrder
    }
}
